### DISCUSSION TITLE
Renovate should not convert commit- and pr-titles to lowercase

### DISCUSSION THREAD

**Discussion by kingcrunch** [renovatebot/renovate#20931] [Suggest an Idea]
### Type of discussion.

I'm proposing an idea

### Tell us more.

Does it makes sense to change the case of the title when semantic-commits-mode is enabled?

While reading the commits, issues, PRs, and so on I thought, that this is really specified somewhere, but now I've read the [Conventional commits spec](https://www.conventionalcommits.org/en/v1.0.0/#specification) and found out, that it is not mentioned there.

Than I found, that only [Angular defined this rule](https://github.com/angular/angular/blob/main/CONTRIBUTING.md#summary). Renovate is used by many projects, even outside the Node/JS/Typescript/... ecosystem, where such rules are unknown in general and at first confusing, because (as mentioned) conventional commits doesn't mention anything like that.

On the other hand, if one likes to have lowercase first character, it is just the [`commitMessageAction`](https://docs.renovatebot.com/configuration-options/#commitmessageaction), which needs a change from "Update" to "update". At least, when there no other, deeper configuration changes I don't see yet :) And even if, I cannot imagine, that is is impossible to setup oneself.

My suggestion:

- Do not change the case of the PR- and commit-title at all and
- Change the default of `commitMessageAction` to `update` (maybe for semantic commits only, but actually I don't know, why to differ)

What do you think?

**Comment by rarkins** [comment]
I'm ok with this idea, but a couple of concerns:
1. I think right now we "detect" angular, not conventional. I'm not sure if we need to enhance our detection capabilities
2. `Update` is probably not the only commitMessageAction affected, but maybe we could solve all of them in a preset

**Reply by kingcrunch** [reply]
I started with a simple PR

Regarding:

1. We enable [`semanticCommit`](https://docs.renovatebot.com/configuration-options/#semanticcommits) manually, so the detection algorithm doesn't affect us. I'll look after that, if I find it ;)
2. I meant, `Update` is the default value. Are there places, were it gets overwritten? Presets probably/maybe?

**Reply by rarkins** [reply]
https://github.com/search?q=repo%3Arenovatebot%2Frenovate%20commitMessageAction&type=code

And specifically:

<img width="513" alt="image" src="https://user-images.githubusercontent.com/6311784/225274628-0476bc3f-9851-4671-a8a7-046ab31ac252.png">

**Comment by rarkins** [comment]
@KingCrunch Let's discuss requirements here again rather than in an active PR.

We have a way of detecting semantic commits, but we don't differentiate between angular and "conventional".

The angular spec says:

```
<type>(<scope>): <short summary>
  │       │             │
  │       │             └─⫸ Summary in present tense. Not capitalized. No period at the end.
  │       │
  │       └─⫸ Commit Scope: animations|bazel|benchpress|common|compiler|compiler-cli|core|
  │                          elements|forms|http|language-service|localize|platform-browser|
  │                          platform-browser-dynamic|platform-server|router|service-worker|
  │                          upgrade|zone.js|packaging|changelog|docs-infra|migrations|
  │                          devtools
  │
  └─⫸ Commit Type: build|ci|docs|feat|fix|perf|refactor|test
```

In other words, the full commit message should be lowercased.

Conventional commits says:
<img width="590" alt="image" src="https://user-images.githubusercontent.com/6311784/226088331-23ba5c73-0828-417c-a48c-a685198600a7.png">

Apart from this, it is the only reference I see to terms like "capitalized", "upper" or "lower" cased except for `BREAKING CHANGES`.

What this says to me is that:
- if angular, lower case the whole thing
- if conventional, don't lower case anything that's not already

Let's say we remove all commit message lower casing, probably in a major release to flag it's a breaking change. In this case probably most people are happy or disinterested, but we also should have a way to allow angular convention users to do what they need (hello @devversion).

I'm pretty sure we don't need to be "angular convention aware" if the only thing is lower casing. We do have a [lowercase template helper](https://docs.renovatebot.com/templates/#lowercase) already but not sure if that's flexible enough to do the *whole* commit message. I suspect it doesn't work as needed when the message itself is *already* a template.

So then a quick solution would be a new option `commitMessageLowerCase` which defaults to `false`, and angular convention users can set it to `true`? @KingCrunch @devversion would this satisfy both your use cases? @KingCrunch would need to do nothing on the next major release while @devversion would need to add this setting as part of the upgrade.

One more thing: we'd need to think about both of these:
- If we are looking for pre-existing closed (ignored) PRs, we should compare them case-insensitive
- If deciding if an *open* PR already exists for a branch, we should match it case-insensitive but detect and update it if the casing has changed

**Reply by kingcrunch** [reply]
I'd be fine with a new option. Maybe (I don't like to promise anything :D) I can look at it later the day.


Just to point it out:

Angular describes it more detailed in their [CONTRIBUTING.md](https://github.com/angular/angular/blob/8d99ad0a39d4092407972265dbd69b4d3ae1b241/CONTRIBUTING.md), which says

> Do not capitalize _the first letter_

![grafik](https://user-images.githubusercontent.com/90388/226104688-2fd021e3-3e69-4ad2-963b-1461a5fa9873.png)

So making the entire subject lowercase would technically introduce a new convention. ;) (subset of angular, which itself is a subset of conventional). Maybe @devversion can clarify. The `CONTRIBUTING.md` might be misleading here.

On the other hand the difference between angular and conventional is _now_ just a single character... So maybe just set the default `messageAction`s (first word of the message after the prefix) lowercase for _all_ users might be still an option? As far as I can see that would already satisfy angulars style.

- `semanticCommits=true` would have now lowercase characters in their title (package names, ...). it would still be angular-compatible. There were issues with lowercased packaged names anyway, don't?
- `semanticCommits=false` would have now the first letter in lowercase. If somebody even cares, that is still configurable back to old style (maybe via preset).

Agree, that this seems to be not solvable without braking change.

**Reply by rarkins** [reply]
> * `semanticCommits=false` would have now the first letter in lowercase. If somebody even cares, that is still configurable back to old style (maybe via preset).

This would be a bad idea, we should not do it

**Reply by kingcrunch** [reply]
As mentioned: I am also fine with the extra option.

> If we are looking for pre-existing closed (ignored) PRs, we should compare them case-insensitive

I haven't found, where this happens yet.

> If deciding if an open PR already exists for a branch, we should match it case-insensitive but detect and update it if the casing has changed

That is already the case, isn't?

https://github.com/renovatebot/renovate/blob/main/lib/workers/repository/update/pr/index.ts#L338-L346

**Reply by kingcrunch** [reply]
I am not sure, if a decision was made.
I've updated the PR-branch with what I think covers the suggestions from @rarkins, the PR is still closed though.

**Reply by devversion** [reply]
Hey all, I haven't read through all discussion here, but from an Angular perspective, it's only important that the **first letter** after the commit scope is not capitalized. 

e.g. `feat(bla): This is wrong, changing EventEmitter`, while `feat(bla): this is ok, changing EventEmitter`.

Generally I've been captializing words/symbols in the "short summary" as needed, when it corresponds to an actual known export or similar. Lower casing the full commit message is not strictly necessary. I don't know what Renovate currently does- but we are also not necessarily bothering a lot I think.

**Reply by kingcrunch** [reply]
I've updated the branch with the suggestions, but now I realize, that github doesn't update the PR now that it is closed

https://github.com/renovatebot/renovate/pull/20930/files

Changes are here

https://github.com/KingCrunch/renovate/tree/fix/do-not-lowercase-entire-commit-title

**Reply by viceice** [reply]
> I am not sure, if a decision was made.
> I've updated the PR-branch with what I think covers the suggestions from @rarkins, the PR is still closed though.
> 

please reopen the PR now

**Reply by kingcrunch** [reply]
@rarkins Can you please reopen the PR?

**Reply by kingcrunch** [reply]
I guess the `breaking`-flag is also not needed anymore now

**Comment by kingcrunch** [comment]
> This would be a bad idea, we should not do it

Could you maybe explain, why? 

To clarify: My suggestion only covers changes in `commitMessageAction`s. What you cited is just, what happens in case of `semanticCommits=false`. The changes would only cover default values, which everybody can revert (manually via options, or presets). There is no logic, that suddenly changes the case of characters anywhere. That is something I try to avoid/reduce, because right now there is some undocumented magic, that makes everything lowercase.

To summarize:

- Remove every "make things magically lowercase"
- Just set `commitMessageAction` to first letter lowercase

Thats it.

Now I dont't understand, why this is worse, than an yet-another-option, that leads to at the end 2 new "conventions" (normal, lowercase-normal, semantic/conventional, lowercase-semantic/"angular").

PS: Don't get me wrong. I just don't get it. If you prefer the extra option, fine. :)

**Reply by rarkins** [reply]
Your suggestion would make millions of future PRs worse by default (missing capitalization). It's a bad user experience if everybody who doesn't use semantic commits has to then change their config to undo this worse behavior. It's important that we have empathy for the users it would affect.

**Comment by rarkins** [comment]
Here's a quick possibility:

New config option `commitMessageLowerCase` which defaults to `"auto"` and has same behavior as today (so that it's backwards compatible).

Also has second option `"never"` which can be configured to disable lowercasing for @KingCrunch.

In a future major release we can:
- Add another option `"first"` which lower cases the first letter only, for Angular usage, and
- Change `"auto"` behavior to be same as `"never"`

**Reply by kingcrunch** [reply]
@rarkins 

Is this your last decision? Because you mentioned another solution, which I implemented (or at least I tried to, because first-time-typescripter as mentioned).

![grafik](https://user-images.githubusercontent.com/90388/226518233-bf30a1c2-0e08-465d-9b8b-7ee4990945e6.png)

I just want to to get this fixed some day...

and we _really_ need this fix, because right now this bug prevents automerging from working properly.

**Reply by rarkins** [reply]
Yes, what I suggested today would allow it to be fixed _now_, without requiring an immediate major release, and gives us more flexibility in future than true/false

**Reply by kingcrunch** [reply]
OK, thanks :)

I've updated the branch
https://github.com/KingCrunch/renovate/tree/fix/do-not-lowercase-entire-commit-title
